### PR TITLE
Refactor UserMenu component.

### PR DESCRIPTION
**Pull request by kostyalmm**
## Summary

Addresses issue #2192 

## Relevant technical choices

* Refactor UserMenu `UserMenu.js` component
* Add `user-menu.stories.js` under `Global`
* Add tests in `UserMenu.test.js`

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/components/UserMenu.js [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	// Log the user out if they confirm the dialog.
+	const handleUnlinkConfirm = useCallback( async () => {
+		// Close the modal.
+		toggleDialog( false );
+
+		// Clear caches.
+		clearWebStorage();
+
+		// Navigate back to the splash screen to reconnect.
+		document.location = getSiteKitAdminURL(
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.js:119]: Let's get this URL from the core/site datastore by calling the `getAdminURL` selector with appropriate arguments.

File: stories/user-menu.stories.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import { STORE_NAME as CORE_SITE } from '../assets/js/googlesitekit/datastore/site/constants';
+import { STORE_NAME as CORE_USER } from '../assets/js/googlesitekit/datastore/user/constants';
+import UserMenu from '../assets/js/components/UserMenu';
+
+storiesOf( 'Global', module )
+	.add( 'UserMenu', () => {
+		const setupRegistry = ( { dispatch } ) => {
+			dispatch( CORE_SITE ).receiveSiteInfo( {
+				proxyPermissionsURL: 'https://sitekit.withgoogle.com/site-management/permissions/',
+			} );
```
Comment (eugene-manuilov) [inline stories/user-menu.stories.js:37]: Let's use `provideSiteInfo` helper.

File: stories/user-menu.stories.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+			dispatch( CORE_USER ).receiveUserInfo( {
+				email: 'test@example.com',
+				picture: 'http://www.gravatar.com/avatar',
+			} );
```
Comment (eugene-manuilov) [inline stories/user-menu.stories.js:41]: Same here, use `provideUserInfo` helper.

File: assets/js/components/UserMenu.test.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import { STORE_NAME as CORE_USER } from '../googlesitekit/datastore/user/constants';
+
+describe( 'UserMenu', () => {
+	let registry;
+
+	beforeAll( () => {
+		registry = createTestRegistry();
+		registry.dispatch( CORE_SITE ).receiveSiteInfo( {
+			proxyPermissionsURL: 'https://sitekit.withgoogle.com/site-management/permissions/',
+		} );
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.test.js:34]: Let's use `provideSiteInfo` helper.

File: assets/js/components/UserMenu.test.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		registry.dispatch( CORE_USER ).receiveUserInfo( {
+			email: 'test@example.com',
+			picture: 'http://www.gravatar.com/avatar',
+		} );
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.test.js:38]: Use `provideUserInfo` helper.

File: assets/js/components/UserMenu.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+			) ) {
+			switch ( index ) {
+				case 0:
+					handleDialog();
+					break;
+				case 1:
+					if ( ! proxyPermissionsURL ) {
+						return;
+					}
+					global.location.assign( proxyPermissionsURL );
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.js:95]:
```suggestion
					if ( proxyPermissionsURL ) {
						global.location.assign( proxyPermissionsURL );
					}
```

File: assets/js/components/UserMenu.js [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	}, [ dialogActive ] );
+
+	const handleMenuItemSelect = useCallback( ( index, e ) => {
+		if (
+			( ( 'keydown' === e.type && (
+				13 === e.keyCode || // Enter
+					32 === e.keyCode // Space
+			) ) ||
+				'click' === e.type // Mouse
+			) ) {
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.js:86]: Let's make is slightly easier to read:

```suggestion
		if (
			( 'keydown' === e.type && ( 13 === e.keyCode || 32 === e.keyCode ) ) || // Enter or Space is pressed.
			'click' === e.type // Mouse is clicked
		) {
```

File: assets/js/components/UserMenu.js [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+					}
+					global.location.assign( proxyPermissionsURL );
+					break;
+				default:
+					handleMenu();
+			}
+		}
+	}, [ proxyPermissionsURL, handleMenu, handleDialog ] );
+
+	const handleDialogClose = ( e ) => {
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.js:103]: Let's use `useCallback` for this function to memoize it.

File: assets/js/components/UserMenu.js [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+		if ( 27 === e.keyCode ) {
+			toggleDialog( false );
+			toggleMenu( false );
+		}
+	};
+
+	// Log the user out if they confirm the dialog.
+	const handleUnlinkConfirm = useCallback( async () => {
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.js:111]: `async` isn't needed here, is it?

File: assets/js/components/UserMenu.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+			global.removeEventListener( 'keyup', handleMenuClose );
+			global.removeEventListener( 'keyup', handleDialogClose );
+		};
+	}, [] );
+
+	const handleMenu = useCallback( () => {
+		toggleMenu( ! menuOpen );
+	}, [ menuOpen ] );
+
+	const handleMenuClose = ( e ) => {
```
Comment (eugene-manuilov) [inline assets/js/components/UserMenu.js:64]: Let's use `useCallback` for this function as well to memoize it.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
It's a good start, @kostyalmm! Left some comments to address.

File: assets/js/components/UserMenu.js (hunk 8 above)
Comment (kostyalmm) [reply assets/js/components/UserMenu.js:103]: IMHO As explained in this post, using `useCallback` here with `[]` dependency is not worth it and will be less efficient than it already is.
If you still feel that a `useCallback` will be a better thing to do here, I will be happy to make the change 😊

File: assets/js/components/UserMenu.js [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { clearWebStorage } from '../util';
+import Dialog from './dialog';
+import Button from './button';
+import Menu from './menu';
+import Modal from './modal';
```
Comment (felixarntz) [inline assets/js/components/UserMenu.js:33]: This will need to be updated after #2321 is merged.

File: assets/js/components/UserMenu.js (hunk 8 above)
Comment (eugene-manuilov) [reply assets/js/components/UserMenu.js:101]: @kostyalmm i am not concerned about performance here. The reason why I think we need to use `useCallback` here is to make sure that we always have the same reference to the same `handleDialogClose` function, so `removeEventListener` calls in the cleanup callback can properly remove event listeners defined on the component mount.

Comment (kostyalmm) [reply assets/js/components/UserMenu.js:101]: Ahh got it. Thanks, making the change

**Review by eugene-manuilov — DISMISSED** [review]
Looks good to me! Thanks!

File: assets/js/components/UserMenu.js [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+						global.location.assign( proxyPermissionsURL );
+					}
+					break;
+				default:
+					handleMenu();
+			}
+		}
+	}, [ proxyPermissionsURL, handleMenu, handleDialog ] );
+
+	const handleDialogClose = useCallback( ( e ) => {
```
Comment (aaemnnosttv) [inline assets/js/components/UserMenu.js:101]: Please move the definition of this function within the `useEffect` above where it is bound to the event listeners as it isn't used anywhere else and this avoids potential bugs related to dependencies.

File: assets/js/components/UserMenu.js [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+					ref={ menuButtonRef }
+					className="googlesitekit-header__dropdown mdc-button--dropdown"
+					text
+					onClick={ handleMenu }
+					icon={ userPicture
+						? <i className="mdc-button__icon" aria-hidden="true"><img className="mdc-button__icon--image"
+							src={ userPicture }
+							alt={ __( 'User Avatar', 'google-site-kit' ) } /></i>
+						: undefined
+					}
```
Comment (aaemnnosttv) [inline assets/js/components/UserMenu.js:137]: Do these need to be defined all on one line? If not let's format it more like we do elsewhere
```js
icon={
	!! userPicture && (
		<i className="mdc-button__icon" aria-hidden="true">
			<img
				className="mdc-button__icon--image"
				src={ userPicture }
				alt={ __( 'User Avatar', 'google-site-kit' ) }
			/>
		</i>
	)
}
```

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @kostyalmm – just a few things to clean up as well as a few changes to make in the tests. Should be fairly quick to address though – let me know if you have any questions.

File: assets/js/components/UserMenu.test.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+		describe( 'and clicking first menu option', () => {
+			it( 'should open the modal dialog', () => {
+				fireEvent.click( menu.children[ 0 ] );
+				expect( document.querySelector( '.mdc-dialog--open' ) ).toBeTruthy();
+			} );
+
+			it( 'should close the modal dialog after pressing escape key', () => {
+				fireEvent.keyUp( document, { keyCode: 27 } );
+				expect( document.querySelector( '.mdc-dialog--open' ) ).toBeNull();
+			} );
```
[the remark this answers is not in the document]
Comment (kostyalmm) [reply assets/js/components/UserMenu.test.js:66]: Yes this was closing the menu. I have fixed this to close the modal now. Thanks for pointing out 😃

File: assets/js/components/UserMenu.test.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+		} );
```
[the remark this answers is not in the document]
Comment (kostyalmm) [reply assets/js/components/UserMenu.test.js:99]: added

**Review by aaemnnosttv — APPROVED** [review]
LGTM 👍

**Pull request merged**